spanner/benchmark/ycsb.py

```diff
@@ -0,0 +1,258 @@
+
+Usage:
+
+  # Set up instance and load data into database.
+
+  # Set up environment variables. You should use your own credentials and gcloud
+  # project.
+  $ export GOOGLE_APPLICATION_CREDENTIALS=/path/to/credentials.json
+  $ export GCLOUD_PROJECT=gcloud-project-name
+
+  # Run the benchmark.
+  $ python spanner/benchmark/ycsb.py run -P pkb/workloada -p table=usertable \
+    -p cloudspanner.instance=ycsb-542756a4 -p recordcount=5000 \
+    -p operationcount=100 -p cloudspanner.database=ycsb -p num_worker=1
+
+  # To make a package so it can work with PerfKitBenchmarker.
+  $ cd spanner; tar -cvzf ycsb-python.0.0.5.tar.gz benchmark/*
+
+"""
+
+import numpy
+import optparse
+import random
+import string
+import threading
+import timeit
+
+from google.cloud import spanner
+
+
+KEYS = []
+OPERATIONS = ['readproportion', 'updateproportion', 'scanproportion',
+              'insertproportion']
+
+
+def parse_options():
+  """Parses options."""
+  parser = optparse.OptionParser()
+  parser.add_option('-P', '--workload', action='store', dest='workload',
+                    default='', help='The path to a YCSB workload file.')
+  parser.add_option('-p', '--parameter', action='append', dest='parameters',
+                    default=[], help='The key=value pair of parameter.')
+  parser.add_option('-b', '--num_bucket', action='store', type='int',
+                    dest='num_bucket', default=1000,
+                    help='The number of buckets in output.')
+
+  options, args = parser.parse_args()
+
+  parameters = {}
+  parameters['command'] = args[0]
+  parameters['num_bucket'] = options.num_bucket
+
+  for parameter in options.parameters:
+    parts = parameter.strip().split('=')
+    parameters[parts[0]] = parts[1]
+
+  with open(options.workload, 'r') as f:
+    for line in f.readlines():
+      parts = line.split('=')
+      key = parts[0].strip()
+      if key in OPERATIONS:
+        parameters[key] = parts[1].strip()
+
+  return parameters
+
+
+def open_database(parameters):
+  """Opens a database specified by the parameters from parse_options()."""
+  spanner_client = spanner.Client()
+  instance_id = parameters['cloudspanner.instance']
+  instance = spanner_client.instance(instance_id)
+  database_id = parameters['cloudspanner.database']
+  pool = spanner.BurstyPool(int(parameters['num_worker']))
+  database = instance.database(database_id, pool=pool)
+
+  return database
+
+
+def load_keys(database, parameters):
+  """Loads keys from database."""
+  del KEYS[:]
+  results = database.execute_sql('SELECT u.id FROM %s u' % parameters['table'])
+
+  for row in results:
+    KEYS.append(row[0])
+
+
+def read(database, table, key):
+  """Does a single read operation."""
+  with database.snapshot() as snapshot:
+    result = snapshot.execute_sql('SELECT u.* FROM %s u WHERE u.id="%s"' %
+                                  (table, key))
+    for row in result:
+      key = row[0]
+      for i in range(10):
+        field = row[i + 1]
```

Comment:
arel there always be 11 objects in row?

Reply:
Yes. The schema is fixed. But let me make a constant for this.